File: Classes/model/TKColoredRoute.m
//
//  AdHocShape.m
//  TripPlanner
//
//  Created by Adrian Schoenig on 21/03/13.
//
//

#import "TKColoredRoute.h"

@interface TKColoredRoute ()

@property (nonatomic, strong) NSArray *routePath;
@property (nonatomic, strong) UIColor *routeColour;
@property (nonatomic, strong) NSArray *routeDashPattern;
@property (nonatomic, assign) BOOL routeIsTravelled;

@end

@implementation TKColoredRoute

- (id)initWithWaypoints:(NSArray *)waypoints
							withColor:(UIColor *)color
						dashPattern:(NSArray *)dashPattern
            isTravelled:(BOOL)isTravelled
{
	self = [super init];
	if (self) {
		self.routePath = waypoints;
		self.routeColour = color;
		self.routeDashPattern = dashPattern;
    self.routeIsTravelled = isTravelled;
	}
	return self;
}

- (id)initWithWaypoints:(NSArray *)waypoints
									 from:(NSInteger)start
										 to:(NSInteger)end
							withColor:(UIColor *)color
						dashPattern:(NSArray *)dashPattern
            isTravelled:(BOOL)isTravelled
{
  NSInteger last = end > 0 ? end : waypoints.count;
  NSRange range = NSMakeRange(start, last - start);
  NSArray *selectedWaypoints = [waypoints subarrayWithRange:range];

  return [self initWithWaypoints:selectedWaypoints
                       withColor:color
                     dashPattern:dashPattern
                     isTravelled:isTravelled];
}

@end

File: Classes/model/TKColoredRoute.swift
//
//  TKColoredRoute.swift
//  TripKit
//
//  Created by Adrian Schoenig on 4/7/17.
//  Copyright © 2017 SkedGo. All rights reserved.
//

import Foundation

public class TKColoredRoute: NSObject {
  
  public let path: [MKAnnotation]
  public let routeColor: SGKColor?
  public let routeDashPattern: [NSNumber]?
  public let routeIsTravelled: Bool
  
  public init(path: [MKAnnotation], color: SGKColor?, dashPattern: [NSNumber]?, isTravelled: Bool) {
    self.path = path
    routeColor = color
    routeDashPattern = dashPattern
    routeIsTravelled = isTravelled
  }

  @objc(initWithWaypoints:from:to:withColor:dashPattern:isTravelled:)
  public init(path: [MKAnnotation], from: Int, to: Int, color: SGKColor?, dashPattern: [NSNumber]?, isTravelled: Bool) {
    let first = from > to ? 0 : from
    let last  = to < from ? path.count - 1 : to
    self.path = Array(path[first...last])
    routeColor = color
    routeDashPattern = dashPattern
    routeIsTravelled = isTravelled
  }
  
}

extension TKColoredRoute: STKDisplayableRoute {
  
  public var routePath: [Any] {
    return path
  }
  
  public var showRoute: Bool { return true }
  
}

